### PR TITLE
CSHARP-2504: BsonWriter: _childElementNameValidatorFactory removal

### PR DESCRIPTION
**Pull request by BorisDog**

### CODE REVIEW

File: src/MongoDB.Bson/IO/BsonWriter.cs [hunk 1]
```diff
@@ -27,8 +27,8 @@ namespace MongoDB.Bson.IO
     public abstract class BsonWriter : IBsonWriter
     {
         // private fields
-        private Func<IElementNameValidator> _childElementNameValidatorFactory = () => NoOpElementNameValidator.Instance;
         private bool _disposed = false;
+        private bool _useChildValidator = false;
```
Comment (rstam) [inline src/MongoDB.Bson/IO/BsonWriter.cs:31]: These changes from `_childElementNameValidatorFactory` to `_useChildValidator` look correct to me.

Nice refactoring.

**Review by rstam — APPROVED** [review]
I'd like someone else to code review the tests (I didn't review them).

LGTM

File: tests/MongoDB.Bson.Tests/IO/BsonWriterTests.cs [hunk 2]
```diff
@@ -41,5 +45,146 @@ public void TestWriteNameThrowsWhenValueIsNull()
                 Assert.Throws<ArgumentNullException>(() => { bsonWriter.WriteName(null); });
             }
         }
+
+        [Theory]
+        [ParameterAttributeData]
+        public void TestValidateName_validnames_should_validate(
```
[the remark this answers is not in the document]
Comment (DmitryLukyanov) [reply tests/MongoDB.Bson.Tests/IO/BsonWriterTests.cs:51]: also, we use a different naming approach for tests. For example: https://github.com/mongodb/mongo-csharp-driver/blob/master/tests/MongoDB.Driver.Tests/EncryptionTests.cs#L159

`BsonWriter_should_use_nameValidator_in_nested_levels` ?

**Comment by BorisDog** [comment]
[EG](https://evergreen.mongodb.com/version/5ff62daaa4cf473235241e66) [Internal Link]

File: tests/MongoDB.Bson.Tests/IO/BsonWriterTests.cs [hunk 3]
```diff
@@ -41,5 +45,146 @@ public void TestWriteNameThrowsWhenValueIsNull()
...
+
+            return result;
+        }
+
+        private Mock<IElementNameValidator> CreatePrefixValidatorMock(string prefix, IElementNameValidator childValidator = null)
+        {
+            var mockValidator = new Mock<IElementNameValidator>();
+
+            mockValidator.Setup(m => m.GetValidatorForChildContent(It.IsAny<string>())).
+                Returns(childValidator ?? NoOpElementNameValidator.Instance);
```
[the remark this answers is not in the document]
Comment (BorisDog) [reply tests/MongoDB.Bson.Tests/IO/BsonWriterTests.cs]: Done.

File: tests/MongoDB.Bson.Tests/IO/BsonWriterTests.cs [hunk 4]
```diff
@@ -41,5 +45,146 @@ public void TestWriteNameThrowsWhenValueIsNull()
...
+            bool addValidatorOnRoot)
+        {
+            var levelNames = Enumerable.Range(0, levelsCount).Select(i => $"level{i}").ToArray();
+            var validatorsMocks = CreatePrefixValidatorsNestedMock(levelNames);
+
+            using (var stream = new MemoryStream())
+            using (var writer = new BsonBinaryWriter(stream))
+            {
+                if (addValidatorOnRoot)
+                    writer.PushElementNameValidator(validatorsMocks[0].Object);
```
[the remark this answers is not in the document]
Comment (BorisDog) [reply tests/MongoDB.Bson.Tests/IO/BsonWriterTests.cs:108]: Done.

File: tests/MongoDB.Bson.Tests/IO/BsonWriterTests.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (BorisDog) [reply tests/MongoDB.Bson.Tests/IO/BsonWriterTests.cs]: Done.

File: tests/MongoDB.Bson.Tests/IO/BsonWriterTests.cs [hunk 5]
```diff
@@ -41,5 +45,146 @@ public void TestWriteNameThrowsWhenValueIsNull()
...
+                for (int i = 0; i < levelsCount; i++)
+                {
+                    writer.WriteEndDocument(); // $"{levelName}_nested"
+                }
+
+                writer.WriteEndDocument();
+            }
+        }
+
+        private Mock<IElementNameValidator>[] CreatePrefixValidatorsNestedMock(string[] prefixes)
```
[the remark this answers is not in the document]
Comment (BorisDog) [reply tests/MongoDB.Bson.Tests/IO/BsonWriterTests.cs:170]: Done.

**Review by DmitryLukyanov — APPROVED** [review]
LGTM

**Pull request merged**